src/memory.rs

```diff
@@ -311,52 +311,70 @@ impl ObjectStore for InMemory {
-        let root = Path::default();
-        let prefix = prefix.unwrap_or(&root);
+    fn list_opts(
+        &self,
+        prefix: Option<&Path>,
+        options: ListOpts,
+    ) -> BoxStream<'static, Result<ListResult>> {
+        if options.delimiter {
+            self.list_with_delimiter(prefix, options.offset.as_ref(), options.max_keys)
+        } else {
+            self.list_without_delimiter(prefix, options.offset.as_ref(), options.max_keys)
+        }
+    }
 
-        let storage = self.storage.read();
-        let values: Vec<_> = storage
-            .map
-            .range((prefix)..)
-            .take_while(|(key, _)| key.as_ref().starts_with(prefix.as_ref()))
-            .filter(|(key, _)| {
-                // Don't return for exact prefix match
-                key.prefix_match(prefix)
-                    .map(|mut x| x.next().is_some())
-                    .unwrap_or(false)
-            })
-            .map(|(key, value)| {
-                Ok(ObjectMeta {
-                    location: key.clone(),
-                    last_modified: value.last_modified,
-                    size: value.data.len() as u64,
-                    e_tag: Some(value.e_tag.to_string()),
-                    version: None,
-                })
-            })
-            .collect();
+    async fn copy(&self, from: &Path, to: &Path) -> Result<()> {
+        let entry = self.entry(from)?;
+        self.storage
+            .write()
+            .insert(to, entry.data, entry.attributes);
+        Ok(())
+    }
 
-        futures::stream::iter(values).boxed()
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> Result<()> {
+        let entry = self.entry(from)?;
+        let mut storage = self.storage.write();
+        if storage.map.contains_key(to) {
+            return Err(Error::AlreadyExists {
+                path: to.to_string(),
+            }
+            .into());
+        }
+        storage.insert(to, entry.data, entry.attributes);
+        Ok(())
     }
+}
 
-    /// The memory implementation returns all results, as opposed to the cloud
-    /// versions which limit their results to 1k or more because of API
-    /// limitations.
-    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> Result<ListResult> {
+impl InMemory {
+    fn list_with_delimiter(
+        &self,
+        prefix: Option<&Path>,
+        offset: Option<&Path>,
+        max_keys: Option<usize>,
+    ) -> BoxStream<'static, Result<ListResult>> {
         let root = Path::default();
         let prefix = prefix.unwrap_or(&root);
+        let offset = offset.unwrap_or(&root);
 
         let mut common_prefixes = BTreeSet::new();
-
         // Only objects in this base level should be returned in the
         // response. Otherwise, we just collect the common prefixes.
         let mut objects = vec![];
         for (k, v) in self.storage.read().map.range((prefix)..) {
+            if let Some(may_keys) = max_keys {
```

Comment:
You could lift the conditional out by doing `let max_keys = max_keys.unwrap_or(usize::MAX)`.